src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -533,15 +533,15 @@ impl<'a, 'b, 'gcx, 'tcx> TypeVerifier<'a, 'b, 'gcx, 'tcx> {
                     }
                 }
                 ty::TyGenerator(def_id, substs, _) => {
-                    // Try upvars first. `field_tys` requires final optimized MIR.
-                    if let Some(ty) = substs.upvar_tys(def_id, tcx).nth(field.index()) {
+                    // Try upvars first. `field_tys` requires the final optimized MIR
+                    if let Some(ty) = substs.pre_transforms_tys(def_id, tcx).nth(field.index()) {
                         return Ok(ty);
                     }
 
                     return match substs.field_tys(def_id, tcx).nth(field.index()) {
```

Comment:
That said, this code is basically dead code, right? That is, we don't run the type-checker after optimization? I guess it's here just in case we do, as a kind of sanity check? (But in that case, i'm not sure that the full set of type checks would pass...)

Reply:
When I wrote it, only the `InstCombine` pass broke type checking. I disabled that and ran MIR type checking later anyway, for debugging purposes.

Reply:
K